src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs

```diff
@@ -0,0 +1,208 @@
+                }
+
+                return null;
+            }
+
+            public override BoundNode VisitQueryClause(BoundQueryClause node)
+            {
+                Visit(node.UnoptimizedForm ?? node.Value);
+                return null;
+            }
+
+            public override BoundNode VisitUnboundLambda(UnboundLambda node)
+            {
+                Visit(node.BindForErrorRecovery().Body);
+                return null;
+            }
+
+            public override BoundNode VisitForEachStatement(BoundForEachStatement node)
+            {
+                this.Visit(node.IterationVariableType);
+                this.Visit(node.Expression);
+                // PROTOTYPE(nullable-api): handle the deconstruction
+                //this.Visit(node.DeconstructionOpt);
+                this.Visit(node.Body);
+                return null;
+            }
+
+            public override BoundNode VisitGotoStatement(BoundGotoStatement node)
+            {
+                // There's no need to verify the label children. They do not have types or nullabilities
+                return null;
+            }
+
+            public override BoundNode VisitTypeOrValueExpression(BoundTypeOrValueExpression node)
+            {
+                Visit(node.Data.ValueExpression);
+                return base.VisitTypeOrValueExpression(node);
+            }
+
+            public override BoundNode VisitDynamicCollectionElementInitializer(BoundDynamicCollectionElementInitializer node)
+            {
+                // https://github.com/dotnet/roslyn/issues/33441 dynamic collection initializers aren't being handled correctly
+                VerifyNode(node, overrideSkippedExpression: true);
+                return null;
+            }
+
+            public override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+            {
+                VisitBinaryOperatorChildren(node);
+                return null;
+            }
+
+            public override BoundNode VisitUserDefinedConditionalLogicalOperator(BoundUserDefinedConditionalLogicalOperator node)
+            {
+                VisitBinaryOperatorChildren(node);
+                return null;
+            }
+
+            private void VisitBinaryOperatorChildren(BoundBinaryOperatorBase node)
+            {
+                // There can be deep recursion on the left side, so verify iteratively to avoid blowing the stack
+                // PROTOTYPE(nullable-api): Handle in rewriter as well
+                while (true)
+                {
+                    VerifyNode(node);
+
+                    Visit(node.Right);
+
+                    if (!(node.Left is BoundBinaryOperatorBase child))
+                    {
+                        Visit(node.Left);
+                        return;
+                    }
+
+                    node = child;
+                }
+            }
+
+            private void VerifyNode(BoundNode node, bool overrideSkippedExpression = false)
```

Comment:
>VerifyNode [](start = 25, length = 10)

nit: consider moving this method to the top of the file (before `Visit` methods)